### PR TITLE
Upgrade plugins to sbt 1.X (Fix #915)

### LINKED ISSUE #915: Upgrade plugins to support sbt 1.0.0
Sbt has added support in 0.13.16 to allow compiling plugins for sbt 1.0.0 to allow easier migration. Getting the plugins ready could smooth the transition to sbt 1.0 and Scala 2.12 usage in plugins.

Release notes:
https://github.com/sbt/sbt/releases/tag/v0.13.16-RC1

Some APIs used in plugins may be private in sbt 1.0. Here is one such solution recommended by the sbt maintainers.
https://github.com/sbt/sbt-release/pull/207/commits/314a2b6bdf9993e65d93424f4ccb444cf13b9638

### PR DESCRIPTION
**Pull request by MasseGuillaume**

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -224,7 +233,7 @@ lazy val tools =
         "com.lihaoyi"    %% "scalaparse" % "0.4.2",
         "org.scalacheck" %% "scalacheck" % "1.13.4" % "test",
         compilerPlugin(
-          "org.scalamacros" % "paradise" % "2.0.1" cross CrossVersion.full),
+          "org.scalamacros" % "paradise" % "2.1.0" cross CrossVersion.full),
```
Comment (MasseGuillaume) [inline build.sbt]: Only 2.1.0 is available for both 2.12.3 and 2.10.6

https://repo1.maven.org/maven2/org/scalamacros/paradise_2.12.3/
https://repo1.maven.org/maven2/org/scalamacros/paradise_2.10.6/

File: sbt-scala-native/src/main/scala/scala/scalanative/sbtplugin/ScalaNativePluginInternal.scala [hunk 2]
```diff
@@ -434,24 +438,31 @@ object ScalaNativePluginInternal {
...
-      val globals = fullClasspath.value
-        .collect { case p if p.data.exists => p.data }
-        .flatMap(p => tools.LinkerPath(VirtualDirectory.real(p)).globals.toSeq)
-
-      globals.map(_.show).sorted
-    },
-    nativeExternalDependencies := ResourceScope { implicit scope =>
+    nativeAvailableDependencies := Def
+      .task {
+        val fcp = fullClasspath.value
```
Comment (MasseGuillaume) [inline sbt-scala-native/src/main/scala/scala/scalanative/sbtplugin/ScalaNativePluginInternal.scala]: This is now an error in sbt.

It's to guard us from the bogus sbt evaluation model. For example:

```
val foo =
  if(false) task1.value
  else task2.value
```

both tasks will be evaluated.

**Comment by MasseGuillaume** [comment]
To select a sbt version you do: `"^^ 1.0.2"`

**Comment by MasseGuillaume** [comment]
@ekrich indeed, updated

**Comment by ekrich** [comment]
So the idea is to support both `0.13.x` and `1.0.x`? Is there a default?

File: build.sbt (hunk 1 above)
Comment (densh) [reply build.sbt]: @xeno-by Is there a reason why macro paradise `2.1.1` is not published for `2.10.6` ?

File: nir/src/main/scala/scala/scalanative/nir/serialization/BinaryDeserializer.scala [hunk 3]
```diff
@@ -100,6 +100,7 @@ final class BinaryDeserializer(_buffer: => ByteBuffer) {
         case T.PinIfAttr =>
           deps += Dep.Conditional(getGlobalNoDep, getGlobalNoDep)
       }
+      ()
```
Comment (densh) [inline nir/src/main/scala/scala/scalanative/nir/serialization/BinaryDeserializer.scala]: Is this change necessary?

Comment (MasseGuillaume) [reply nir/src/main/scala/scala/scalanative/nir/serialization/BinaryDeserializer.scala]: Yes it does not compile in Scala 2.12

**Comment by densh** [comment]
@ekrich Given that `1.0` is the future of sbt you should at least try it out. But it's still relatively new and `0.13.16` is a safe middle-ground until the ecosystem migrates to `1.0`.

**Comment by ekrich** [comment]
@densh Agreed. I was just surprised at all the code it takes to support both.

File: build.sbt (hunk 1 above)
Comment (xeno-by) [reply build.sbt]: Oversight. I'll remedy this shortly.

Comment (MasseGuillaume) [reply build.sbt]: @densh any reason not to bump this, all tests are passing.

File: sbt-scala-native/src/main/scala/scala/scalanative/sbtplugin/ScalaNativePluginInternal.scala [hunk 4]
```diff
@@ -434,24 +438,31 @@ object ScalaNativePluginInternal {
...
+        ResourceScope { implicit scope =>
+          val globals = fcp
+            .collect { case p if p.data.exists => p.data }
+            .flatMap(p =>
+              tools.LinkerPath(VirtualDirectory.real(p)).globals.toSeq)
+
+          globals.map(_.show).sorted
+        }
+      }
+      .dependsOn(compile)
```
Comment (sjrd) [inline sbt-scala-native/src/main/scala/scala/scalanative/sbtplugin/ScalaNativePluginInternal.scala]: `fullClasspath` indirectly depends on `compile` for the same configuration, so this line is unnecessary. Consequently, you don't need the `Def.task` wrapper.

**Comment by MasseGuillaume** [comment]
@Duhemm to run travis with both sbt version can I do the following?

` - bash sbt -J-Xmx2G ^rebuild "set scriptedBufferLog in sbtScalaNative := false" ^test-all`

**Comment by sjrd** [comment]
> I think the best solution for Travis would be to duplicate all the jobs... Using ^, we're certain to run in timeouts on almost every build. Maybe overwrite project/build.properties before the job starts? I think that's what Scala.JS does.

Not exactly. We mostly use `^^` in duplicated jobs. We only overwrite `project/build.properties` inside the sbt-plugin-test directory, which has its own build to test our sbt plugin. Our main build only boots on 0.13.16, then we use `^^` to build the sbt plugin *for* 1.x *on* 0.13. See https://github.com/scala-js/scala-js/blob/bd0e45d00933cce7a774b6f149b1020e76d4287f/ci/matrix.xml#L199-L222 for the gory details.

File: sbt-scala-native/src/main/scala/scala/scalanative/sbtplugin/Utilities.scala [hunk 5]
```diff
@@ -2,16 +2,13 @@ package scala.scalanative
 package sbtplugin
 
 import java.lang.System.{lineSeparator => nl}
+import scala.sys.process.{Process, _}
```
[the remark this answers is not in the document]
Comment (MasseGuillaume) [reply sbt-scala-native/src/main/scala/scala/scalanative/sbtplugin/Utilities.scala]: This clashes on 0.13 with sbt imports.

File: build.sbt [hunk 6]
```diff
@@ -163,7 +162,13 @@ lazy val noPublishSettings = Seq(
 lazy val toolSettings =
   baseSettings ++
     Seq(
-      scalaVersion := toolScalaVersion,
+      crossSbtVersions := List("0.13.16", "1.0.2"),
+      scalaVersion := {
+        (sbtBinaryVersion in pluginCrossBuild).value match {
+          case "0.13" => "2.10.6"
+          case _      => "2.12.3"
```
[the remark this answers is not in the document]
Comment (MasseGuillaume) [reply build.sbt]: I just tried this and it does not work properly.

If I start sbt and do `^^ 1.0.2` then `sbtScalaNative/compile` it wants to fetch sbt-crossproject _2.10.

[the remark this answers is not in the document]
Comment (Duhemm) [reply build.sbt]: Argh sure, I was assuming you'd be starting with either sbt 1 or sbt 0.13. I can't think of anything better than leaving this as it is then.

**Comment by MasseGuillaume** [comment]
It looks like some scripted test need to be migrated: https://travis-ci.org/scala-native/scala-native/jobs/295230751#L8345

File: build.sbt (hunk 1 above)
Comment (xeno-by) [reply build.sbt]: Wow this is the definition of things falling through the cracks. I've published `2.1.1` for `2.10.7`. Would it be fine for you guys to migrate?

Comment (densh) [reply build.sbt]: `2.10.7` should work.

**Comment by MasseGuillaume** [comment]
All the comments are addressed in this PR and the build is green. :100:

**Comment by densh** [comment]
@Duhemm do you have any more comments here?

**Comment by Duhemm** [comment]
@densh Nope, that's all I had.

**Comment by MasseGuillaume** [comment]
Hum the ci stalls on SBT 1.0
master
https://travis-ci.org/scala-native/scala-native/builds/307862335

vs 
It works fine on the CI for this PR:
https://travis-ci.org/scala-native/scala-native/builds/300174813

it's stuck in scripted java-io-file-1

```
Resolving org.scala-sbt#test-interface;1.0 ...
Resolving org.scala-native#nscplugin_2.11.11;0.4.0-SNAPSHOT ...
Resolving org.scala-native#nscplugin_2.11.11;0.4.0-SNAPSHOT ...
Resolving org.scala-lang#scala-compiler;2.11.11 ...
Resolving org.scala-lang.modules#scala-xml_2.11;1.0.5 ...
Resolving org.scala-lang.modules#scala-parser-combinators_2.11;1.0.4 ...
Resolving jline#jline;2.14.3 ...
Done updating.
Compiling 2 Scala sources to /tmp/sbt_b9dd7aeb/java-io-file-1/target/scala-2.11/classes...
Linking (3093 ms)
Discovered 1310 classes and 9619 methods
Optimizing (6627 ms)
Generating intermediate code (1372 ms)
Produced 37 files
Compiling to native code (2752 ms)
Linking native code (109 ms)
Total time: 20 s, completed Nov 27, 2017 1:17:32 PM
Compiling 1 Scala source to /tmp/sbt_b9dd7aeb/java-io-file-1/target/scala-2.11/classes...
Linking (2112 ms)
Discovered 1314 classes and 9646 methods
Optimizing (3076 ms)
Generating intermediate code (1080 ms)
Produced 37 files


No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received

The build has been terminated
```

on this PR the test continued:

```
Compiling to native code (3140 ms)
Linking native code (117 ms)
Total time: 20 s, completed Nov 10, 2017 2:19:42 PM
Compiling 1 Scala source to /tmp/sbt_6db308ef/java-io-file-1/target/scala-2.11/classes...
Linking (2241 ms)
Discovered 1309 classes and 9545 methods
Optimizing (4356 ms)
Generating intermediate code (922 ms)
Produced 37 files
Compiling to native code (2994 ms)
Linking native code (119 ms)
Total time: 12 s, completed Nov 10, 2017 2:19:53 PM
...

**Comment by ekrich** [comment]
@MasseGuillaume I just rebased my changes on top of yours but upgraded to sbt `1.0.4` and `2.12.4` and got a clean build - https://travis-ci.org/scala-native/scala-native/builds/307911322

File: sbt-scala-native/src/main/scala/scala/scalanative/sbtplugin/ScalaNativeCrossVersion.scala [hunk 7]
```diff
@@ -5,38 +5,21 @@ package sbtplugin
...
+  val currentBinaryVersion = binaryVersion(Versions.current)
 
-    def binaryVersion(full: String): String = full match {
-      case ReleaseVersion(major, minor, _) => s"$major.$minor"
-      case _                               => full
-    }
-
-    binaryVersion(Versions.current)
+  def binaryVersion(full: String): String = full match {
+    case ReleaseVersion(major, _, _) => major
```
Comment (densh) [inline sbt-scala-native/src/main/scala/scala/scalanative/sbtplugin/ScalaNativeCrossVersion.scala]: This change would have been devastating if it made into a release. It turned `native0.3` binary version into `native0`. #1096

Comment (MasseGuillaume) [reply sbt-scala-native/src/main/scala/scala/scalanative/sbtplugin/ScalaNativeCrossVersion.scala]: nice catch!
